// @workspace/db/src/schema/artisans.ts
import { pgTable, text, serial, timestamp, integer, real, pgEnum } from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";
import { usersTable } from "./users";

export const verificationStatusEnum = pgEnum("verification_status", ["pending", "approved", "rejected"]);

export const artisanProfilesTable = pgTable("artisan_profiles", {
  id: serial("id").primaryKey(),
  userId: integer("user_id").notNull().references(() => usersTable.id, { onDelete: "cascade" }).unique(),
  bio: text("bio"),
  city: text("city"),
  avatarUrl: text("avatar_url"),
  nationalIdImageUrl: text("national_id_image_url"),
  verificationStatus: verificationStatusEnum("verification_status").notNull().default("pending"),
  completedJobs: integer("completed_jobs").notNull().default(0),
  reviewsCount: integer("reviews_count").notNull().default(0),
  averageRating: real("average_rating"),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow().$onUpdate(() => new Date()),
});

export const specialtiesTable = pgTable("specialties", {
  id: serial("id").primaryKey(),
  name: text("name").notNull().unique(),
});

export const artisanSpecialtiesTable = pgTable("artisan_specialties", {
  id: serial("id").primaryKey(),
  artisanId: integer("artisan_id").notNull().references(() => artisanProfilesTable.id, { onDelete: "cascade" }),
  specialtyId: integer("specialty_id").notNull().references(() => specialtiesTable.id, { onDelete: "cascade" }),
});

export const insertArtisanProfileSchema = createInsertSchema(artisanProfilesTable).omit({ id: true, createdAt: true, updatedAt: true });
export type InsertArtisanProfile = z.infer<typeof insertArtisanProfileSchema>;
export type ArtisanProfile = typeof artisanProfilesTable.$inferSelect;

export const insertSpecialtySchema = createInsertSchema(specialtiesTable).omit({ id: true });
export type InsertSpecialty = z.infer<typeof insertSpecialtySchema>;
export type Specialty = typeof specialtiesTable.$inferSelect;
