// @workspace/db/src/schema/offers.ts
import { pgTable, text, serial, timestamp, integer, real, pgEnum } from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";
import { serviceRequestsTable } from "./requests";
import { artisanProfilesTable } from "./artisans";

export const offerStatusEnum = pgEnum("offer_status", ["pending", "accepted", "rejected", "withdrawn"]);

export const offersTable = pgTable("offers", {
  id: serial("id").primaryKey(),
  requestId: integer("request_id").notNull().references(() => serviceRequestsTable.id, { onDelete: "cascade" }),
  artisanId: integer("artisan_id").notNull().references(() => artisanProfilesTable.id, { onDelete: "cascade" }),
  price: real("price").notNull(),
  arrivalTimeText: text("arrival_time_text").notNull(),
  estimatedDurationText: text("estimated_duration_text").notNull(),
  message: text("message").notNull(),
  status: offerStatusEnum("status").notNull().default("pending"),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow().$onUpdate(() => new Date()),
});

// Negotiation rounds linked to an offer
export const offerNegotiationsTable = pgTable("offer_negotiations", {
  id: serial("id").primaryKey(),
  offerId: integer("offer_id").notNull().references(() => offersTable.id, { onDelete: "cascade" }),
  senderRole: text("sender_role").notNull(), // "customer" | "artisan"
  proposedPrice: real("proposed_price").notNull(),
  message: text("message"),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
});

export const insertOfferSchema = createInsertSchema(offersTable).omit({ id: true, createdAt: true, updatedAt: true });
export type InsertOffer = z.infer<typeof insertOfferSchema>;
export type Offer = typeof offersTable.$inferSelect;
export type OfferNegotiation = typeof offerNegotiationsTable.$inferSelect;
