// @workspace/db/src/schema/requests.ts
import { pgTable, text, serial, timestamp, integer, pgEnum, date } from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";
import { usersTable } from "./users";

export const requestStatusEnum = pgEnum("request_status", [
  "open",
  "receiving_offers",
  "offer_selected",
  "in_progress",
  "completed",
  "cancelled",
]);

export const serviceRequestsTable = pgTable("service_requests", {
  id: serial("id").primaryKey(),
  customerId: integer("customer_id").notNull().references(() => usersTable.id, { onDelete: "cascade" }),
  title: text("title").notNull(),
  category: text("category").notNull(),
  description: text("description").notNull(),
  city: text("city").notNull(),
  locationText: text("location_text").notNull(),
  preferredDate: date("preferred_date").notNull(),
  status: requestStatusEnum("status").notNull().default("open"),
  selectedOfferId: integer("selected_offer_id"),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow().$onUpdate(() => new Date()),
});

export const serviceRequestImagesTable = pgTable("service_request_images", {
  id: serial("id").primaryKey(),
  requestId: integer("request_id").notNull().references(() => serviceRequestsTable.id, { onDelete: "cascade" }),
  imageUrl: text("image_url").notNull(),
});

export const insertServiceRequestSchema = createInsertSchema(serviceRequestsTable).omit({ id: true, createdAt: true, updatedAt: true });
export type InsertServiceRequest = z.infer<typeof insertServiceRequestSchema>;
export type ServiceRequest = typeof serviceRequestsTable.$inferSelect;
